### ISSUE TITLE
Script tests

### ISSUE THREAD

**Issue by libbitc** [jgarzik/picocoin#83]
Hi,

Just to note that some CSV tests that pass in bitcoin-core are failing here so have been removed from `test/script_tests.json` for now. See libbitc/libbitc#2
Also, the automatically generated tests from bitcoin-core `test/script_tests.json`have been removed

I do not think this is a problem with `lib/script_eval.c` but rather `test/script.c` will probably need to be updated for these removed tests to pass.
Once we get over this issue all script tests from bitcoin-core will pass except the SCRIPT_VERIFY_WITNESS and SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_WITNESS_PROGRAM  ones.

**Comment by jgarzik** [comment]
ACK - agree - thanks for filing

**Comment by aido** [comment]
Hi,

`SCRIPT_VERIFY_NULLDUMMY` will also need to be added before all bitcoin-core scriot tests pass:

https://github.com/jgarzik/picocoin/blob/master/include/ccoin/script.h#L55-L56

**Comment by aido** [comment]
Hi,

Two of the failing script tests that have been removed are:
```
["0", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME", "CSV fails if stack top bit 1 << 31 is set and the tx version < 2"],
["4294967296", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME",
  "CSV fails if stack top bit 1 << 31 is not set, and tx version < 2"],
```
These seem to be failing due to a segfault here:
https://github.com/jgarzik/picocoin/blob/master/lib/script_eval.c#L541-L547

When these tests are run `txTo->vin` is NULL, causing the segfault at line 543.

**Comment by libbitc** [comment]
Hi,

Looking at the script test code in bitcoin-core I think the segfault reported above is happening because some tests need a test transaction which is not yet created in picocoin `test/script.c`. So `txTo->vin` is NULL as pointed out.

Here's how bitcoin-core does it: https://github.com/bitcoin/bitcoin/blob/0.12/src/test/script_tests.cpp#L114-L145

Adding something like the following to `test/script.c` should get most of the removed script tests from bitcoin-core to pass in picocoin. I haven't got it working yet though, it still needs some tweaking:
```c
struct bitc_tx BuildCreditingTransaction(struct cstring* scriptPubKey)
{
    struct bitc_tx txCredit;
    bitc_tx_init(&txCredit);
    txCredit.nVersion = 1;
    txCredit.nLockTime = 0;
    txCredit.vin = parr_new(0, bitc_txin_freep);
    txCredit.vout = parr_new(0, bitc_txout_freep);

    struct bitc_txin *txinCredit = calloc(1, sizeof(struct bitc_txin));
    bitc_txin_init(txinCredit);
    bitc_outpt_null(&txinCredit->prevout);
    txinCredit->scriptSig = cstr_new(NULL);
    txinCredit->nSequence = SEQUENCE_FINAL;
    parr_add(txCredit.vin, txinCredit);

    struct bitc_txout *txoutCredit = calloc(1, sizeof(struct bitc_txout));
    bitc_txout_init(txoutCredit);
    txoutCredit->scriptPubKey = cstr_new_buf(scriptPubKey->str, scriptPubKey->len);
    txoutCredit->nValue = 0;
    parr_add(txCredit.vout, txoutCredit);
    bitc_tx_calc_sha256(&txCredit);

    return txCredit;
}

struct bitc_tx BuildSpendingTransaction(struct cstring* scriptSig, struct bitc_tx txCredit)
{
    struct bitc_tx txSpend;
    bitc_tx_init(&txSpend);
    txSpend.nVersion = 1;
    txSpend.nLockTime = 0;
    txSpend.vin = parr_new(0, bitc_txin_freep);
    txSpend.vout = parr_new(0, bitc_txout_freep);

    struct bitc_txin *txinSpend = calloc(1, sizeof(struct bitc_txin));
    bitc_txin_init(txinSpend);
    bu256_copy(&txinSpend->prevout.hash, &txCredit.sha256);
    txinSpend->prevout.n = 0;
    txinSpend->scriptSig = cstr_new_buf(scriptSig->str, scriptSig->len);
    txinSpend->nSequence = SEQUENCE_FINAL;
    parr_add(txSpend.vin, txinSpend);

    struct bitc_txout *txoutSpend = calloc(1, sizeof(struct bitc_txout));
    bitc_txout_init(txoutSpend);
    txoutSpend->scriptPubKey = cstr_new(NULL); //
    txoutSpend->nValue = 0;
    parr_add(txSpend.vout, txoutSpend);

    bitc_tx_free(&txCredit);
    return txSpend;
}

static void test_script(bool is_valid,cstring *scriptSig, cstring *scriptPubKey,
			unsigned int idx, const char *scriptSigEnc,
			const char *scriptPubKeyEnc,
			const unsigned int test_flags)
{
    struct bitc_tx tx = BuildSpendingTransaction(scriptSig, BuildCreditingTransaction(scriptPubKey));;

	bool rc;
	rc = bitc_script_verify(scriptSig, scriptPubKey, &tx, 0,
			      test_flags, SIGHASH_NONE);
.
.
.
```

**Comment by libbitc** [comment]
Hi,

I have fixed the issue where the following script tests fail:
```
["0", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME", "CSV fails if stack top bit 1 << 31 is set and the tx version < 2"],
["4294967296", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME",
  "CSV fails if stack top bit 1 << 31 is not set, and tx version < 2"],
```
A pull request with the fix will follow shortly.

However there are still a few more script tests failing. One in particular is this one:
```
["0 0 1", "EQUAL EQUAL", "P2SH,STRICTENC", "OK", "OP_0 and bools must have identical byte representations"],
```
Unlike bitcoin-core, picocoin does not seem to have identical byte representations for OP_0 and bools.

Again, here's how bitcoin does it:
https://github.com/bitcoin/bitcoin/blob/master/src/script/script.h#L554-L568
and here's how picocoin does it:
https://github.com/jgarzik/picocoin/blob/master/lib/script.c#L108-L112

**Comment by jgarzik** [comment]
@libbitc hmmm interesting... very useful report and in need of further investigation

**Comment by libbitc** [comment]
Hi,

Looking at the bitcoin code:
https://github.com/bitcoin/bitcoin/blob/0.13/src/script/script.h#L379-L388

A picocoin problem may lie around here:
https://github.com/jgarzik/picocoin/blob/master/lib/script.c#L278-L286

Which should possibly look like this?
```c
void bsp_push_int64(cstring *s, int64_t n)
{
	if (n == -1 || (n >= 1 && n <= 16)) {
		cstr_append_c(s, (unsigned char) (n + (OP_1 - 1)));
		return;
	} else if (n == 0) {
		cstr_append_c(s, (unsigned char) (OP_0));
		return;
	}

	bool neg = false;
```
And similar for `bsp_push_uint64()`

But that does not seem to be the only problem causing the OP_0/bools test to fail.

**Comment by aido** [comment]
Hi,

> Once we get over this issue all script tests from bitcoin-core will pass except the SCRIPT_VERIFY_WITNESS and SCRIPT_VERIFY_DISCOURAGE_UPGRADABLE_WITNESS_PROGRAM ones.

I'm not certain but I think that support for P2SH and MULTISIG will need to be added before all script tests from bitcoin-core pass. See `lib/script_sign.c`:
```c
	case TX_SCRIPTHASH:		/* TODO; not supported yet */
	case TX_MULTISIG:
		goto out;
```
This is possibly why most of the automatically generated tests in `test/script_tests.json` had to be removed.

**Comment by libbitc** [comment]
Hi,

Most of the failing/removed script tests are failing because of an issue with OP_CHECKSIG.
OP_CHECKSIG  is failing for a few reasons.

OP_CHECKSIG calls `bp_checksig()`, The following code should be removed from `bp_checksig()`:
https://github.com/jgarzik/picocoin/blob/master/lib/script_eval.c#L310-L316
```c
	/* examine hashtype at end of string, then remove it */
	unsigned char *vch_back = vchSigHT->p + (vchSigHT->len - 1);
	if (nHashType == 0)
		nHashType = *vch_back;
	else if (nHashType != *vch_back)
		return false;
	struct buffer vchSig = { vchSigHT->p, vchSigHT->len - 1 };
```
and replaced with just:
```c
	/* examine hashtype at end of string, then remove it */
	unsigned char *vch_back = vchSigIn->p + (vchSigIn->len - 1);
 	int nHashType = *vch_back;
	struct buffer vchSig = { vchSigIn->p, vchSigIn->len - 1 };
```
Look at https://github.com/bitcoin/bitcoin/blob/master/src/script/interpreter.cpp#L1259-L1264 for inspiration.

Then`bp_checksig()` calls `bp_verify()`. For OP_CHECKSIG operations `bp_verify()` incorrectly returns false.

I would wager that OP_CHECKSIG fails because the scriptCode string is not correct: 
https://github.com/jgarzik/picocoin/blob/master/lib/script_eval.c#L1247-L1255

**Comment by libbitc** [comment]
Hi,

The OP_CHECKSIG tests are failing because as suggested above the output of `bp_tx_sighash()` is incorrect causing `bp_verify()` to fail. The output of `bp_tx_sighash()` is incorrect because during  OP_CHECKSIG tests the hash value in `txin->prevout` is incorrect.

`test/script.c` has two functions `BuildSpendingTransaction()` and `BuildCreditingTransaction()` with the hash value in prevout of the spending transaction being a hash of the crediting transaction. This hash value is incorrect ultimately causing `bp_verify()` to fail during testing,

So the root cause of the problem is that the `BuildCreditingTransaction()` function in `test/script.c` does not create a crediting transaction identical to that created by bitcoin-core.

Comparing bitcoin-core's `BuildCreditingTransaction()` function:
https://github.com/bitcoin/bitcoin/blob/master/src/test/script_tests.cpp#L123-L137

to picocoin's `BuildCreditingTransaction()` function:
https://github.com/jgarzik/picocoin/blob/master/test/script.c#L14-L38

I suspect that the problem lies around how picocoin implements  `txCredit.vin[0].scriptSig = CScript() << CScriptNum(0) << CScriptNum(0);`

**Comment by libbitc** [comment]
Hi @jgarzik,

OP_CHECKSIG test issue now fixed in PR #99.
With the exception of three CLEANSTACK tests below all script tests from bitcoin-core 0.12 (pre-SegWit) are now passing:

```json
[
    "11 0x47 0x304402202f7505132be14872581f35d74b759212d9da40482653f1ffa3116c3294a4a51702206adbf347a2240ca41c66522b1a22a41693610b76a8e7770645dc721d1635854f01 0x43 0x410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ac",
    "HASH160 0x14 0x31edc23bdafda4639e669f89ad6b2318dd79d032 EQUAL",
    "P2SH",
    "OK",
    "P2SH with unnecessary input but no CLEANSTACK"
],
[
    "11 0x47 0x304402202f7505132be14872581f35d74b759212d9da40482653f1ffa3116c3294a4a51702206adbf347a2240ca41c66522b1a22a41693610b76a8e7770645dc721d1635854f01 0x43 0x410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ac",
    "HASH160 0x14 0x31edc23bdafda4639e669f89ad6b2318dd79d032 EQUAL",
    "CLEANSTACK,P2SH",
    "CLEANSTACK",
    "P2SH with unnecessary input"
],
[
    "0x47 0x304402202f7505132be14872581f35d74b759212d9da40482653f1ffa3116c3294a4a51702206adbf347a2240ca41c66522b1a22a41693610b76a8e7770645dc721d1635854f01 0x43 0x410479be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798483ada7726a3c4655da4fbfc0e1108a8fd17b448a68554199c47d08ffb10d4b8ac",
    "HASH160 0x14 0x31edc23bdafda4639e669f89ad6b2318dd79d032 EQUAL",
    "CLEANSTACK,P2SH",
    "OK",
    "P2SH with CLEANSTACK"
],
```

**Comment by jgarzik** [comment]
Closing as fixed after merging #99
